benchmarks/humaneval.py

```diff
@@ -0,0 +1,205 @@
+"""
+HumanEval: Hand-Written Evaluation Set
+
+Mark Chen, Jerry Tworek, Heewoo Jun, Qiming Yuan, Henrique Ponde de Oliveira Pinto,
+Jared Kaplan, Harri Edwards, Yuri Burda, Nicholas Joseph, Greg Brockman, Alex Ray,
+Raul Puri, Gretchen Krueger, Michael Petrov, Heidy Khlaaf, Girish Sastry,
+Pamela Mishkin, Brooke Chan, Scott Gray, Nick Ryder, Mikhail Pavlov, Alethea Power,
+Lukasz Kaiser, Mohammad Bavarian, Clemens Winter, Philippe Tillet,
+Felipe Petroski Such, Dave Cummings, Matthias Plappert, Fotios Chantzis,
+Elizabeth Barnes, Ariel Herbert-Voss, William Hebgen Guss, Alex Nichol, Alex Paino,
+Nikolas Tezak, Jie Tang, Igor Babuschkin, Suchir Balaji, Shantanu Jain,
+William Saunders, Christopher Hesse, Andrew N. Carr, Jan Leike, Josh Achiam,
+Vedant Misra, Evan Morikawa, Alec Radford, Matthew Knight, Miles Brundage,
+Mira Murati, Katie Mayer, Peter Welinder, Bob McGrew, Dario Amodei, Sam McCandlish,
+Ilya Sutskever, Wojciech Zaremba
+https://arxiv.org/pdf/2107.03374
+
+Based on: https://github.com/openai/simple-evals/blob/main/humaneval_eval.py
+"""
+
+import re
+from collections import defaultdict
+
+import numpy as np
+
+from inspect_ai import Task, task
+from inspect_ai.dataset import Sample, hf_dataset
+from inspect_ai.scorer import (
+    CORRECT,
+    INCORRECT,
+    Metric,
+    Score,
+    Scorer,
+    Target,
+    metric,
+    scorer,
+)
+from inspect_ai.solver import Generate, TaskState, generate, solver
+from inspect_ai.util import subprocess
+
+INSTRUCTION = "Read the following function signature and docstring, and fully implement the function described. Your response should only contain the code for this function.\n"
+TIMEOUT = 20
+NUM_EPOCHS = 5
+
+
+def _per_sample_collator(scores: list[Score]) -> dict[str, list[float]]:
```

Comment:
I might not be getting the full gist of this, but is this doing some per-epoch aggregation? Should this be pulled into Inspect as a first class feature? If that would be helpful let's discuss in more depth.

Reply:
Yes, that’s correct. HumanEval does reductions across epochs, then across samples. IIUC, the default behaviour in Inspect is to treat all samples across all epochs as one large dataset to reduce over. 

I wouldn’t be surprised if other benchmarks require a similar strategy so we’re happy to discuss ways to extend the functionality within Inspect!